### PR TITLE
PLT-64: plutus-conformance: tests for all error cases of evaluator

### PR DESCRIPTION
**Pull request by bezirg**
Adds more tests that exercice different cases the machines can fail.
I created one uplc example for every different kind of cek machine error or cek user error.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting master unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-conformance/test-cases/uplc/evaluation/failure/cons-bs-arg-overflow/cons-bs-arg-overflow.uplc [hunk 1]
```diff
@@ -0,0 +1,2 @@
+-- the arg overflow'ed over the maxBound :: Word8
+(program 1.0.0 [(builtin consByteString) (con integer 256) (con bytestring #)])
```
Comment (michaelpj) [inline plutus-conformance/test-cases/uplc/evaluation/failure/cons-bs-arg-overflow/cons-bs-arg-overflow.uplc:2]: I think we have a specific builtins folder for behaviour of specific builtins, which I think this is?

Comment (michaelpj) [reply plutus-conformance/test-cases/uplc/evaluation/failure/cons-bs-arg-overflow/cons-bs-arg-overflow.uplc:2]: and some more below

File: plutus-conformance/test-cases/uplc/evaluation/failure/head/head.uplc [hunk 2]
```diff
@@ -0,0 +1 @@
+(program 0.0.0 [(force (builtin headList)) [(builtin mkNilData) (con unit ())]])
```
Comment (michaelpj) [inline plutus-conformance/test-cases/uplc/evaluation/failure/head/head.uplc:1]: comment?

**Review by michaelpj — COMMENTED** [review]
This looks like good coverage of builtin errors, are we covering all the machine stuff? e.g. do we have applying a non-function to something?

File: plutus-conformance/test-cases/uplc/evaluation/failure/cons-bs-arg-overflow/cons-bs-arg-overflow.uplc (hunk 1 above)
Comment (thealmarty) [reply plutus-conformance/test-cases/uplc/evaluation/failure/cons-bs-arg-overflow/cons-bs-arg-overflow.uplc:2]: so we have a `builtins` folder and a `failure` folder. Not sure if it's a sensible organization? But if it is, then failures should all go in the `failure` folders.

Comment (michaelpj) [reply plutus-conformance/test-cases/uplc/evaluation/failure/cons-bs-arg-overflow/cons-bs-arg-overflow.uplc:2]: I'm not sure that "failure" is such an interesting categorization. I think people are more likely to be interested in a class of behaviour, and then they're going to be interested in both successful and failing behaviour. so my money goes for "just put them all in the builtins folder"

**Comment by bezirg** [comment]
> This looks like good coverage of builtin errors, are we covering all the machine stuff? e.g. do we have applying a non-function to something?

argExpected.uplc

File: plutus-conformance/test-cases/uplc/evaluation/failure/cons-bs-arg-overflow/cons-bs-arg-overflow.uplc (hunk 1 above)
Comment (thealmarty) [reply plutus-conformance/test-cases/uplc/evaluation/failure/cons-bs-arg-overflow/cons-bs-arg-overflow.uplc:2]: Sounds good to me.

**Comment by bezirg** [comment]
@thealmarty can you check one last time?

**Comment by thealmarty** [comment]
`consByteString` still failing due to Agda version as you pointed out. We need to fix it or add it to the expected failure list for now.

Could you clarify a bit more what we have tested on `verifyEcdsaSecp256k1Signature`? I think you included some of this info in the file names but did we test all of:
- invalid key (including key too long or too short or just invalid or just empty)
- invalid key with edge case messages (empty or just spaces or weird things like that)

**Comment by bezirg** [comment]
> `consByteString` still failing due to Agda version as you pointed out. We need to fix it or add it to the expected failure list for now.
Thanks for that i fixed it i think.

> Could you clarify a bit more what we have tested on `verifyEcdsaSecp256k1Signature`? I think you included some of this info in the file names but did we test all of:
> 
> * invalid key (including key too long or too short or just invalid or just empty)
> * invalid key with edge case messages (empty or just spaces or weird things like that)

I only tested the `verify*Signature` builtins for things that can lead to (evaluation failure). For this to happen, the msg/key/sig has to have some format error. I did not add tests that exercise if the cek succees but the verification of the signature returns `False`.

**Comment by bezirg** [comment]
> > `consByteString` still failing due to Agda version as you pointed out. We need to fix it or add it to the expected failure list for now.
> 
> Thanks for that i fixed it i think.
> 
> > Could you clarify a bit more what we have tested on `verifyEcdsaSecp256k1Signature`? I think you included some of this info in the file names but did we test all of:
> > 
> > * invalid key (including key too long or too short or just invalid or just empty)
> > * invalid key with edge case messages (empty or just spaces or weird things like that)
> 
> I only tested the `verify*Signature` builtins for things that can lead to (evaluation failure). For this to happen, the msg/key/sig has to have some format error. I did not add tests that exercise if the cek succeeds with the result of verification of the signature returning `False`.

In other words, I did not test for "invalid key with edge case messages (empty or just spaces or weird things like that)"

**Review by thealmarty — APPROVED** [review]
